### ISSUE TITLE
Add App Extension API Only Framework

### ISSUE THREAD

**Issue by bassrock** [rs/SDWebImage#1082]
With the latest master SDWebImage, WebImage iOS 8 Framework, you can not use this in Extension targets. There should be also be an app extension safe version.

**Comment by flovilmart** [comment]
@bassrock what's the problem with App Extensions? Could we only have 1 framework target that works all round instead of 2 targets?

**Comment by bassrock** [comment]
@flovilmart There are about 4 places where SDWebImage calls sharedApplication for background tasks. sharedApplication can not be used in app extensions, like today views, watch extensions, or other frameworks.

I made it a seperate target because I did not know of any other way to do this properly.

**Comment by flovilmart** [comment]
Ok, that make sense.

We should check at runtime for sharedApplication so we have a single framework that work in every case

**Comment by bassrock** [comment]
Will that work for Carthage still?

**Comment by flovilmart** [comment]
@bassrock that should not interfere with Carthage as the only thing Carthage needs to build the framework is the shared scheme. 
I'd rather to keep everything under one roof instead of having multiple targets if that makes sense for you.

@bpoplauschi what do you think?

**Comment by flovilmart** [comment]
Instead of adding the flag we could check for:

    Class UIApplicationClass = NSClassFromString(@"UIApplication");
    if(!UIApplicationClass || ![UIApplicationClass respondsToSelector:@selector(sharedApplication)]) {
        return;
    }

**Comment by flovilmart** [comment]
Check it out there: flovilmart/SDWebImage/tree/AppExtensionSafe

**Comment by bassrock** [comment]
@flovilmart so that won't work, when you check the app extension safe box in Xcode for the framework a build will complain because it checks at compile time.

**Comment by flovilmart** [comment]
I don't see the issue here

**Comment by flovilmart** [comment]
@bassrock with carthage the frameworks are built with the carthage tool so your project doesn't build the frameworks itself. In that case, Xcode can't complaint because it actually never builds the framework.

**Comment by bassrock** [comment]
While true you still get a warning about linking dylib unless the App Extension was checked on the framework and my understanding is that if we were to go with your proposed solution the framework won't build with that checked, and its possible you can get a rejection. I could be totally wrong

**Comment by flovilmart** [comment]
The solution I propose check at runtime the availability of UIApplication and sharedApplication, that can't lead to a rejection. And that disables the UIApplication related features, the same way your compile time flag does.

**Comment by bassrock** [comment]
@flovilmart I just downloaded your branch and tried to build it with the "Allow App extension API only" checked. And get a build failure.

 
![screen shot 2015-03-20 at 3 24 06 pm](https://cloud.githubusercontent.com/assets/1010384/6761908/4b5fb160-cf15-11e4-88dc-155790f2eaae.png)

**Comment by flovilmart** [comment]
Ok that makes sense, Do you have treat warnings as error and “Require Only App-Extension-Safe API” build setting to Yes?

I,ve set the Require Only App-Extension-Safe API to NO and it runs properly
as you can see there:
![ios simulator screen shot 2015-03-20 18 33 05](https://cloud.githubusercontent.com/assets/364568/6761975/90ba86b2-cf2f-11e4-83b7-02665892ddd7.png)

**Comment by bassrock** [comment]
Yes it does run properly when you disable that checkbox. However you still get a linking against dylib is unsafe when that is unchecked, which could still trigger a rejection.

"To configure an app extension target to use an embedded framework, set the target’s “Require Only App-Extension-Safe API” build setting to Yes. If you don’t, Xcode reminds you to do so by displaying the warning “linking against dylib not safe for use in application extensions”. "

https://developer.apple.com/library/prerelease/ios/documentation/General/Conceptual/ExtensibilityPG/ExtensionScenarios.html

**Comment by bassrock** [comment]
However based on that and not having tried submitting an app with that warning I am unsure if it actually will trigger a rejection. But I do not doubt it.

**Comment by flovilmart** [comment]
I've updated my branch with performSelector:@selector(sharedApplication) and the warning goes away, as well as the warning at link time with Require Only App-Extension-Safe-API =  YES

**Comment by bassrock** [comment]
Ok that sounds better then! Sweet! Now my question would be why did AFNetworking do the compile flags vs. this way. More so curiosity/specific reason for submission or something.

**Comment by flovilmart** [comment]
I really dunno you should ask them :)

The performSelector method is kinda dirty as it’s runtime manipulation

**Comment by ruslanskorb** [comment]
@flovilmart The method that you suggested is not correct, since the class `UIApplication` is available in App Extension and it responds to the method `sharedApplication`. But we have no right to use it because Apple will reject an application that will use this method.

To resolve this problem, you can use my [category](https://github.com/ruslanskorb/RSKImageCropper/blob/master/RSKImageCropper/UIApplication%2BRSKImageCropper.m) and simple condition:
```obj-c
UIApplication *application = [UIApplication rsk_sharedApplication];
if (application) {
    // call any method of the app instance
}
```

**Comment by flovilmart** [comment]
Did you got real life approved apps with that strategy? I don't mind which one we use.

**Comment by ruslanskorb** [comment]
@flovilmart Not yet, but I don't see any reason that could cause rejection.

**Comment by flovilmart** [comment]
In both cases it's tomatos and tomatos.

Both check for availability of selector at runtime. Your implementation is pretty much the same in the end as the repondsToSelector calls the runtime underneath.

**Comment by bassrock** [comment]
I will be trying with @flovilmart's method to submit an app within the next few days with WatchKit and TodayView support. So far it works with Apple's TestFlight so we will see if it can make it past a real review.

**Comment by flovilmart** [comment]
You,re taking a bullet for the team!


I believe the restriction is meant to prevent any calls to sharedApplication from watch kit as it may crash the glance or the extension.




I don’t think having the code there would actually lead to a rejection, still I may be wrong  :)




On another note, the check is kinda messy and may be wrapped into a [UIApplication safeSharedApplication]

**Comment by bassrock** [comment]
@flovilmart yea we will see. With the floodgates opened this morning it's now just deciding if it is ready.

**Comment by bassrock** [comment]
@flovilmart well my app went into review this morning using this patch. We should have an answer soon.

**Comment by flovilmart** [comment]
Awesome!

**Comment by justinschier** [comment]
@bassrock @flovilmart Any update on this? Was your app accepted?

**Comment by bassrock** [comment]
@justinschier yes it was. I thought i replied to this but it looks like my reply got deleted.

**Comment by flovilmart** [comment]
Hi @justinschier , it's been a while... We met at Gamification Conference at SF back 3 years ago!

**Comment by gwk** [comment]
Hello, I am wrestling with this issue as well. I'm curious what the implications are of skipping the backgroundCleanDisk calls; does this mean that the cache just grows forever, or is there something that extension code must do manually to trim the cache?

**Comment by cwhstcnn** [comment]
@bassrock It's nice that your app is accepted. I can use this library again, hahh~ 
@flovilmart what's the meaning of "xxx that disables the UIApplication related features"? Can you tell me something about the related features? THX a lot...

**Comment by flovilmart** [comment]
@cwhstcnn basically when in app extensions, Apple forbids calling to [UIApplication sharedApplication] as it doesn't really make sense. The [UIApplication sharedApplication] is the host app, and an extension can't really call any methods on it as it's likely not being in the same memory space. 

So what I did was make sure that features relying on UIApplication were properly not called at runtime but adding the runtime checks. Mostly it removes the backgroundTask management provided by the application when using SDWebImage into an app extension.

@gwk I believe Applications and their extensions share the same folders (tmp, documents etc...) (a per documentation) so in all logic, they should share the same cache; therefore, the backgroundCleanDisk won't be skipped by the main application and cache will be properly cleaned. (Still I haven't checked)

**Comment by gwk** [comment]
@flovilmart thanks; that was what we settled on as well.

**Comment by cwhstcnn** [comment]
@flovilmart  THX,,,it really helps me a lot

**Comment by bpoplauschi** [comment]
Fixed with #1085. @flovilmart @bassrock @auibrian @mythodeia many thanks for working on this. For 4.0, we will take a deeper look at #1182 and see if we can switch to that.
